src/main/scala/it/unibo/authsim/library/dsl/otp/builders/OTPBuilder.scala

```diff
@@ -0,0 +1,119 @@
+import it.unibo.authsim.library.dsl.policy.builders.StringPoliciesBuilders.OTPPolicyBuilder
+import it.unibo.authsim.library.dsl.policy.model.StringPolicies.OTPPolicy
+
+import java.util.concurrent.TimeUnit
+import scala.concurrent.duration.Duration
+import scala.language.postfixOps
+import scala.util.Random
+
+/**
+ * ''OTPBuilder'' is a trait that is used to build an new one time password of the type T
+ * @tparam T the type to build
+ */
+trait OTPBuilder[T] extends Builder[T]:
+  /**
+   * Set a [[OTPPolicy one time password policy]]
+   * @param policy otp policy to set
+   * @param generateLength a generator of the actual length of one time password given a otp policy
+   * @return instance of the actual builder
+   */
+  def withPolicy(policy: OTPPolicy)(implicit generateLength: OTPPolicy => Int): this.type
+
+object OTPBuilder:
+  /**
+   * It rappresents the user's userID and password pair.
+   */
+  type SecretValue = (String, String)
+
+  /**
+   *  ''SecretBuilder'' rappresents an extension builder to use Secret value.
+   */
+  trait SecretBuilder:
+    /**
+     * Set the secret key (@see [[SecretValue]])
+     * @param secret key to set
+     * @return instance of the actual builder
+     */
+    def secret(secret: SecretValue): this.type
+
+  /**
+   *  ''HmacOTPBuilder'' rappresents an extension builder to implement HOTP.
+   */
+  trait HmacOTPBuilder:
+    /**
+     * Set the [[HashFunction hash function]] used to generate the otp.
+     * @param hashFunction hash function to set
+     * @return instance of the actual builder
+     */
+    def hashFunction(hashFunction: HashFunction): this.type
+
+  /**
+   * ''TimeOTPBuilder'' rappresents an extension builder to implement TOTP.
+   */
+  trait TimeOTPBuilder:
+    /**
+     * Set the [[Duration time]] that rappresents the expiration of one time password
+     * @param timeout time to set
+     * @return instance of the actual builder
+     */
+    def timeout(timeout: Duration): this.type
+
+  /**
+   * ''AbstractOTPBuilder'' is an abstract builder that implements [[OTPBuilder]] methods.
+   *
+   * Default:
+   *  - [[OTPPolicy OTP policy]] have a minimum length of 4 numbers and a maximum of 10 numbers.
+   *  - [[SecretValue Secret value]] was a random string of 10 characters.
+   *
+   * @tparam T  the type to build
+   */
+  abstract class AbstractOTPBuilder[T] extends OTPBuilder[T] with SecretBuilder:
+    protected val SEPARATOR_SECRET = '-'
+    protected var _policy: OTPPolicy = null
+    protected var _secret: String = Random.alphanumeric.take(10).mkString
+
+    protected var _length: Int = 0
+    this.withPolicy(OTPPolicyBuilder() minimumLength 4 maximumLength 10 build)
+
+    protected var _seed: Int = 0
+    this.generateSeed
+
+    protected def generateSeed(implicit seedGenerator: () => Int): Unit =
+      val genSeed: Int = seedGenerator()
+      this._seed = if genSeed == this._seed then genSeed++ else genSeed
```

Comment:
In this line, `genSeed++` and `genSeed` return the same value; are you sure it has the intended behavior?

If possible, the update of `genSeed` should be moved to a separate line to explicit the mutation (and separate it from the assignment of `this._seed`

Reply:
geenSeed++ and genSeed are not the same : geenSeed++  is genSeed + 1. I could remove ++ and just write genSeed + 1.
Why separate them? Isn't it more compact like this?

Reply:
The separation was suggested as I thought `genSeed++` had the same meaning it has in Java or C.

>  I could remove `++` and just write `genSeed + 1`.

If it is the only place `++` is used, it is better.